File: leetcode/python/flatten_binary_tree_to_linked_list/flatten_binary_tree_to_linked_list.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return nothing, do it in place
    def flatten(self, root):
         if root is None:
             return
         node_list = []
         self.preorder(root,node_list)
         for i,node in enumerate(node_list):
             if i < len(node_list) - 1:
                 node.right = node_list[i+1]
                 node.left = None
             else:
                 node.left = None
                 node.right = None
         
    def preorder(self,root,node_list):
        if root is None:
            return
        node_list.append(root)
        self.preorder(root.left,node_list)
        self.preorder(root.right,node_list)
        

File: leetcode/cpp/FlattenBinaryTreetoLinkedList.cpp
//
// Created by zhangqi on 2018/1/12.
//

#include "TreeNode.h"

class Solution {
public:
    void flatten(TreeNode* root) {
        if(!root){
            return;
        }
        if(root->left){
            flatten(root->left);
        }
        if(root->right){
            flatten(root->right);
        }

        TreeNode *tmp = root->right;
        root->right = root->left;
        root->left = nullptr;
        while (root->right){
            root = root->right;
        }
        root->right = tmp;
    }
};